Add WENO limiters

fmahebert opened the pull request:
## Proposed changes

Add implementations of,
- the Zhong _et al_ (2013) simple WENO limiter
- the Zhu _et al_ (2016) Hermite WENO limiter

Depends on
- [x] #1540 
- [x] #1554 

The last two commits are new.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Add WenoType enumeration

fmahebert committed:
Add WENO limiters

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoType.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#include <ostream>
+
+class Option;
```
`/// \cond` around the forward decls please

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoType.hpp:
```diff
@@ -0,0 +1,32 @@
+enum class WenoType { Hweno, SimpleWeno };
+
+std::ostream& operator<<(std::ostream& os,
+                         const SlopeLimiters::WenoType& weno_type);
```
`noexcept`

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoType.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "Options/ParseOptions.hpp"
+
+std::ostream& SlopeLimiters::operator<<(
+    std::ostream& os, const SlopeLimiters::WenoType& weno_type) {
```
`noexcept`

Since it's an enum, taking by value is probably better

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+  using options = tmpl::list<Type, NeighborWeight, DisableForDebugging>;
+  static constexpr OptionString help = {"A WENO limiter for DG"};
+
+  explicit Weno(const WenoType weno_type, const double neighbor_linear_weight,
```
does not need to be `explicit` since only the last argument of 3 is defaulted.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+
+  explicit Weno(const WenoType weno_type, const double neighbor_linear_weight,
+                const bool disable_for_debugging = false) noexcept
+      : weno_type_(weno_type),
```
the general suggestion is not to have constructors be inline since they're surprisingly expensive

fmahebert replied:
For my education: you mean expensive to compile?

nilsdeppe replied:
Yes, compile time :)

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+  ~Weno() = default;
+
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) noexcept {  // NOLINT
```
maybe we can move this out of the class body too?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+                                           ::Tags::SizeOfElement<VolumeDim>>;
+
+  /// \brief Package data for sending to neighbor elements
+  void package_data(const gsl::not_null<PackagedData*>& packaged_data,
```
This function is very large. Could you move it out of the class body?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+                 ::Tags::SizeOfElement<VolumeDim>>;
+
+  /// \brief Limit the solution on the element
+  bool operator()(
```
this function is very large. Could we move it out of the class body?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+///
+/// To reconstruct the WENO solution from the local solution and the modified
+/// neighbor solutions, the standard WENO procedure is followed. We use the
+/// oscillation indicator of \cite Dumbser2007, modified for use on the square/
```
[nitpick] could you but `squared/cubed` together. Right now an extra space is being added in the rendered documentation

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+/// To reconstruct the WENO solution from the local solution and the modified
+/// neighbor solutions, the standard WENO procedure is followed. We use the
+/// oscillation indicator of \cite Dumbser2007, modified for use on the square/
+/// cube grids of Spectre. We favor this indicator because portions of the work
```
Spectre -> SpECTRE

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+/// neighbor solutions, the standard WENO procedure is followed. We use the
+/// oscillation indicator of \cite Dumbser2007, modified for use on the square/
+/// cube grids of Spectre. We favor this indicator because portions of the work
+/// can be precomputed, leading to an efficient oscillation measure while using
```
I would find `while` -> `when` to be clearer, but in both cases it seems a bit redundant since I'm not sure when else one uses a limiter :man_shrugging:

fmahebert replied:
Heh, that was definitely an oversight on my part. I'll restructure the sentence.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+#include "NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp"
+#include "NumericalAlgorithms/LinearOperators/MeanValue.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Gsl.hpp"
```
`Utilities/Algorithm.hpp`

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,354 @@
+template <size_t VolumeDim>
+struct Element;
+template <size_t VolumeDim>
+struct Mesh;
```
Since the code that calls WENO doesn't know what files to include for the tags, I think this header should do the include.

fmahebert replied:
To be sure I understand: you're suggesting replacing the two Tags::X forward decls with proper includes?

nilsdeppe replied:
Yes, that's right

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,878 @@
+      test_creation<SlopeLimiters::Weno<1, tmpl::list<ScalarTag>>>(
+          "  Type: SimpleWeno");
+
+  // Test default TVBM value, operator==, and operator!=
```
TVBM value?

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,878 @@
+          "  Type: SimpleWeno");
+
+  // Test default TVBM value, operator==, and operator!=
+  CHECK(hweno_default == hweno_default_mixing);
```
Could you also compare to a not-from-options constructed class?

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,878 @@
+  CHECK(hweno_default != hweno_more_mixing);
+  CHECK(hweno_default != simple_weno_default);
+
+  test_creation<SlopeLimiters::Weno<1, tmpl::list<ScalarTag>>>("  Type: Hweno");
```
Could you test that the result is constructed properly?

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,878 @@
+        neighbor_data,
+    const VariablesMap<VolumeDim>& expected_neighbor_modified_vars,
+    // NOLINTNEXTLINE(google-runtime-references):
+    Approx& local_approx = approx) noexcept {
```
It's not clear to me why this needs to be a mutable reference. Could you explain?

fmahebert replied:
The approx doesn't want to be const. I found some other instances in the code that take it by mutable reference so I went with that... but it could also be passed by value.

nilsdeppe replied:
Awe :( Taking by value seems somewhat cleaner so we don't need the `NOLINT` but :man_shrugging:

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Weno.cpp:
```diff
@@ -0,0 +1,878 @@
+    Approx& local_approx = approx) noexcept {
+  // First run some sanity checks on the input, and make sure the test function
+  // is being called in a reasonable way
+  ASSERT(element.neighbors().size() == neighbor_data.size(),
```
Should all these `ASSERT`s not be `ERROR`s?

nilsdeppe reviewed (commented):
I think I reviewed the right commits (GitHub has them out of order)

Overall, looks great, just some minor comments. I like the design a lot :+1: Good job! :)

nilsdeppe commented:
Could you rebase this on develop now that all dependencies have been merged?

I'm also happy for you to squash things related to my comments :)

fmahebert commented:
Rebased; changes addressing @nilsdeppe's review are squashed in.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoType.cpp:
```diff
@@ -0,0 +1,44 @@
+  static constexpr OptionString help = {"Example help text"};
+};
+
+void check_weno_type_parse(const std::string& input,
```
Could you try the `test_creation` function to see if it'll simplify this code a bit? It's in `tests/Unit/TestCreation.hpp`

fmahebert replied:
I am not having success here.

As far as I can tell, the only difference between my function `check_weno_type_parser` and `test_creation` is the additional `\n` character that `test_creation` inserts between the dummy class name and the input string. If I understand things correctly, this means that `test_creation` constructs classes that contain options, whereas my function constructs the option directly. Is my understanding correct?

fmahebert replied:
If my understanding is correct, one path forward would be to add a new `test_enum_creation` or similar that could be used here, in MinmodType (which has an essentially identical `check_minmod_type_parse`), and wherever else may be useful...?

nilsdeppe replied:
Okay, I'm fine with leaving as is for now or creating the `test_enum_creation` now and filing an issue :) Whichever you prefer

nilsdeppe commented on docs/References.bib:
```diff
@@ -604,6 +616,19 @@ @article{Yee1999
+@article{Zhong2013,
+  author   = "Zhong, Xinghui and Shu, Chi-Wang",
+  title    = "A simple weighted essentially nonoscillatory limiter for
+              Runge–Kutta discontinuous {Galerkin} methods",
```
Please also put `Runge-Kutta` in braces so it's capitalized correctly

fmahebert replied:
For my education: is `{Runge}-{Kutta}` or `{Runge-Kutta}` the preferred way of doing this?

kidder replied:
I don't think there is a preferred way, all caps within the `{}`s are preserved.  I've seen bitex with just the individual letters enclosed (e.g. `{R}unge-{K}utta`) or the entire string...  Looks like the existing Zhu reference does `{Runge-Kutta}` so I'd go with that...

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,373 @@
+/// solution when the solution is smooth. Full volume data is communicated
+/// between neighbors.
+///
+/// The limiter uses a Minmod-based troubled-cell indicator to indentify
```
`indentify` -> `identify`

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,373 @@
+/// "modified" solution from each neighbor element. For the simple WENO limiter,
+/// the modified solution is obtained by simply extrapolating the neighbor
+/// solution onto the troubled element. For the HWENO limiter, the modified
+/// solution is obtained by a least-squared fit to the solution across multiple
```
`least-squared` -> `least-squares`

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,373 @@
+                                           ::Tags::SizeOfElement<VolumeDim>>;
+
+  /// \brief Package data for sending to neighbor elements
+  void package_data(const gsl::not_null<PackagedData*>& packaged_data,
```
We can just take the `not_null` by value since it's a pointer

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Weno.hpp:
```diff
@@ -0,0 +1,373 @@
+
+  /// \brief Limit the solution on the element
+  bool operator()(
+      const gsl::not_null<std::add_pointer_t<db::item_type<Tags>>>... tensors,
```
no need to `const` in the forward decl the `not_null`

fmahebert replied:
The compiler disagrees.

nilsdeppe replied:
Huh, well I don't know the rules for parameter packs but that is strange... :man_shrugging:

nilsdeppe reviewed (changes requested):
Group review with @wthrowe and @kidder 

Please squash the changes immediately :+1:

fmahebert commented:
Squashed in a few fixes.

It wasn't clear to me how to address the comment about using `test_creation`. If it turns out to be non-trivial, I would suggest revisiting this in a future PR.

fmahebert committed:
Add test_enum_creation test helper

fmahebert committed:
Use test_enum_creation to test MinmodType

fmahebert committed:
Use test_enum_creation to test WenoType

fmahebert commented:
- Added `test_enum_creation` to `TestCreation.hpp`, then changed `MinmodType` and `WenoType` tests to use this new helper.
- Rebased on `develop` for PDALs 🚲

nilsdeppe reviewed (approved):
LGTM!

fmahebert commented:
I've tried restarting the gcc-8 release build several times but it keeps timing out..

nilsdeppe commented:
My guess is that a PR we merged while this build was in progress caused cache misses and therefore the timeout. I see all builds were already restarted, which is what I was going to suggest :)
